addons/mail/controllers/discuss.py

```diff
@@ -573,3 +573,29 @@ def mail_guest_update_name(self, guest_id, name):
         if guest_to_rename_sudo != guest and not request.env.user._is_admin():
             raise NotFound()
         guest_to_rename_sudo._update_name(name)
+
+    # --------------------------------------------------------------------------
+    # Link preview API
+    # --------------------------------------------------------------------------
+
+    @http.route('/mail/link_preview', methods=['POST'], type='json', auth='public')
+    def mail_link_preview(self, message_id, urls):
+        throttle = request.env['ir.config_parameter'].sudo().get_param('mail.link_preview_throttle', 99)
+        if int(throttle) == 0:
+            return [('clear',)]
+        link_previews = []
+        guest = request.env['mail.guest']._get_guest_from_request(request)
+        for url in set(urls[0:5]):
+            link_preview = request.env['mail.link.preview'].sudo().with_context(guest=guest).create_link_preview(url, message_id)
+            if link_preview:
+                link_previews.append(('insert', link_preview._link_preview_format()))
+        return link_previews or [('clear',)]
```

Comment:
Instead of an array of insert, this could be an insert-and-replace of array
then clear is not necessary, replace empty array is valid